### ISSUE TITLE
不能超过 245 字节

### ISSUE THREAD

**Issue by yangzhencheng** [stulzq/DotnetCore.RSA#2]
return Convert.ToBase64String(_publicKeyRsaProvider.Encrypt(Encoding.UTF8.GetBytes(text), RSAEncryptionPadding.Pkcs1));

在运行时， _publicKeyRsaProvider.Encrypt 报告 Encoding.UTF8.GetBytes(text) 不能超过 245

系统输出：
CryptographicException: The message exceeds the maximum allowable length for the chosen options (245).

**Comment by IgorCova** [comment]
RSA is only able to encrypt data to a maximum amount of your key size (2048 bits = 256 bytes) minus padding/header data (11 bytes for PKCS#1 v1.5 padding).

As a result, it is often not possible to encrypt files with RSA directly. RSA is also not meant for this purpose. If you want to encrypt more data, you can use [Hybrid cryptosystem](https://en.wikipedia.org/wiki/Hybrid_cryptosystem) 

Scenario:
1. All of your data you encrypt with any secure symmetric cipher (e.g. AES, 3DES, Salsa ...)
2. Your key from symmetric cipher you have to encrypt with public key RSA
3. Send your encrypted data with encrypted keys

In another side somebody got your message at first they decrypt key with private key RSA and then with the decrypted key they will be able to decrypt data 

Reasons behind this approach are:
1. RSA encryption is SLOW. Very slow compared to symmetric one.
2. Even for small output, the RSA encryption output is way longer
3. RSA in certain situations (small message space) may have some weaknesses

Basically, you create a random encryption key, encrypt data of any length using any secure symmetric cipher (e.g. AES, 3DES, Salsa, whatever..) and then use RSA to encrypt the symmetric encryption key.

**Comment by stulzq** [comment]
RSA算法本来就不能加密过大的文本，如有需要请自行分段加密

**Comment by yangzhencheng** [comment]
> RSA is only able to encrypt data to a maximum amount of your key size (2048 bits = 256 bytes) minus padding/header data (11 bytes for PKCS#1 v1.5 padding).
> 
> As a result, it is often not possible to encrypt files with RSA directly. RSA is also not meant for this purpose. If you want to encrypt more data, you can use [Hybrid cryptosystem](https://en.wikipedia.org/wiki/Hybrid_cryptosystem)
> 
> Scenario:
> 
> 1. All of your data you encrypt with any secure symmetric cipher (e.g. AES, 3DES, Salsa ...)
> 2. Your key from symmetric cipher you have to encrypt with public key RSA
> 3. Send your encrypted data with encrypted keys
> 
> In another side somebody got your message at first they decrypt key with private key RSA and then with the decrypted key they will be able to decrypt data
> 
> Reasons behind this approach are:
> 
> 1. RSA encryption is SLOW. Very slow compared to symmetric one.
> 2. Even for small output, the RSA encryption output is way longer
> 3. RSA in certain situations (small message space) may have some weaknesses
> 
> Basically, you create a random encryption key, encrypt data of any length using any secure symmetric cipher (e.g. AES, 3DES, Salsa, whatever..) and then use RSA to encrypt the symmetric encryption key.

Thx :) 我对加密这方面的了解不是很多，十分感谢您的解答

**Comment by yangzhencheng** [comment]
> RSA算法本来就不能加密过大的文本，如有需要请自行分段加密

十分感谢，因为对加密算法了解不多。我会多多读相关的资料，给您添麻烦了。感恩！

**Comment by IgorCova** [comment]
> > RSA算法本来就不能加密过大的文本，如有需要请自行分段加密
> 
> 十分感谢，因为对加密算法了解不多。我会多多读相关的资料，给您添麻烦了。感恩！

感恩节！
